libmamba/src/core/history.cpp

```diff
@@ -325,4 +325,183 @@ namespace mamba
+        // removed: -conda-forge/linux-64::xtl-0.8.0-h84d6215_0 mamba2 broken format: installed:
+        // +conda-forge::xtl-0.8.0-h84d6215_0 removed:
+        // -https://conda.anaconda.org/conda-forge/linux-64::xtl-0.8.0-h84d6215_0 mamba2 new format:
+        // installed: +https://conda.anaconda.org/conda-forge/linux-64::xtl-0.8.0-h84d6215_0
+        // removed: -https://conda.anaconda.org/conda-forge/linux-64::xtl-0.8.0-h84d6215_0
+        specs::PackageInfo pkg_info_builder(std::string s)
+        {
+            std::string s_pkg;
+            std::string channel;
+            std::size_t pos_0 = s.rfind("/");
+            if (pos_0 != std::string::npos)
+            {
+                std::string s_begin = s.substr(0, pos_0);
+                std::string s_end = s.substr(pos_0 + 1, s.size());
+                std::size_t pos = s_begin.rfind("/");
+                if (pos != std::string::npos)
+                {
+                    channel = s_begin.substr(pos + 1, s_begin.size());
+                }
+                else
+                {
+                    channel = s_begin;
+                }
+                s_pkg = std::get<1>(util::rsplit_once(s_end, "::"));
+            }
+            else
+            {
+                channel = std::get<0>(util::split_once(s, "::"));
+                s_pkg = std::get<1>(util::rsplit_once(s, "::"));
+            }
+
+            std::size_t pos_1 = s_pkg.rfind("-");
+            std::string s_pkg_ = s_pkg.substr(0, pos_1);
+            std::string build_string = s_pkg.substr(pos_1 + 1, s_pkg.size());
+
+            std::size_t pos_2 = s_pkg_.rfind("-");
+            std::string name = s_pkg_.substr(0, pos_2);
+            std::string version = s_pkg_.substr(pos_2 + 1, s_pkg_.size());
+
+            specs::PackageInfo pkg_info{ name, version, build_string, channel };
+            return pkg_info;
+        }
+
+        PackageDiff PackageDiff::get_revision_pkg_diff(
+            std::vector<History::UserRequest> user_requests,
+            int target_revision
+        )
+        {
+            struct revision
+            {
+                int key = -1;
+                std::map<std::string, specs::PackageInfo> removed_pkg = {};
+                std::map<std::string, specs::PackageInfo> installed_pkg = {};
+            };
+
+            std::vector<revision> revisions;
+            for (auto r : user_requests)
+            {
+                if ((r.link_dists.size() > 0) || (r.unlink_dists.size() > 0))
+                {
+                    if (r.revision_num > target_revision)
+                    {
+                        revision rev{ /*.key = */ r.revision_num };
+                        for (auto ud : r.unlink_dists)
+                        {
+                            auto pkg_info = pkg_info_builder(ud);
+                            auto name = pkg_info.name;
+                            rev.removed_pkg[name] = std::move(pkg_info);
```

Comment:
Consider using `.insert` instead of `[key]` for inserting as this would overwrite existing key-values . Not sure if that's intended or not. Also `operator[]` will first create an empty element before overwriting after `operator=` so I think most of the time `.insert` (or `.emplace` in different cases) is better.